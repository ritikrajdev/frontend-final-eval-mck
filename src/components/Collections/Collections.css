.collections > h2 {
  padding: 24px 20px;
  background-color: white;
}

.collections .entries {
  display: flex;
  justify-content: space-between;
}

.collections .entries button {
  color: var(--dark-purple);
  font-weight: 600;
  display: flex;
  font-size: 1.2rem;
  justify-content: center;
  align-items: center;
  cursor: pointer;
}

.collections i {
  font-size: 1.2rem;
  cursor: pointer;
}

.entries-table {
  margin-top: 20px;
  display: flex;
  flex-direction: column;
  gap: 20px;
}

.entries-headings {
  display: flex;
  padding: 10px 20px;
}

.w-150 {
  width: 150px;
}

.entries-row {
  display: flex;
  align-items: center;
  background-color: white;
  padding: 10px 20px;
  border-radius: 5px;
}
