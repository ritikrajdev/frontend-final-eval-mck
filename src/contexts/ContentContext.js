import PropTypes from 'prop-types';
import React, { createContext } from 'react';

export const ContentContext = createContext([]);

export default function ContentContextProvider({ children }) {
  const [content, setContent] = React.useState(null);

  return (
    <ContentContext.Provider value={[content, setContent]}>
      {children}
    </ContentContext.Provider>
  );
}

ContentContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
