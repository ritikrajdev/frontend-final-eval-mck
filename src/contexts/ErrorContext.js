import PropTypes from 'prop-types';
import React, { createContext } from 'react';

export const ErrorContext = createContext([]);

export default function ErrorContextProvider({ children }) {
  const [error, setError] = React.useState(null);

  if (error) {
    setTimeout(() => {
      setError(null);
    }, 5000);
  }

  return (
    <ErrorContext.Provider value={[error, setError]}>
      {error && (
        <div
          style={{
            position: 'absolute',
            bottom: '20px',
            left: '20px',
            backgroundColor: 'white',
            display: 'inline-block',
            padding: '20px',
            borderRadius: '8px',
            color: 'red',
            border: '1px solid black',
          }}
        >
          {error}
        </div>
      )}
      {children}
    </ErrorContext.Provider>
  );
}

ErrorContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
