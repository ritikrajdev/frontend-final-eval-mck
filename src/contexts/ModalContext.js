import PropTypes from 'prop-types';
import React, { createContext } from 'react';
import Button from '../components/elements/Button';
import Input from '../components/elements/Input';

export const ModalContext = createContext([]);

export default function ModalContextProvider({ children }) {
  const [modal, setModal] = React.useState({
    // position: 'center' | 'right'
    show: false,
    position: 'center',
    title: 'Modal',
    fields: {},
    onSave: () => {
      return true;
    },
  });

  let positionStylingProps = {};

  if (modal.position === 'center') {
    positionStylingProps = {
      top: '50%',
      left: '50%',
      transform: 'translate(-50%, -50%)',
    };
  } else if (modal.position === 'right') {
    positionStylingProps = {
      top: '0',
      right: '0',
      bottom: '0',
    };
  }

  return (
    <ModalContext.Provider value={[modal, setModal]}>
      {modal.show && (
        <div
          style={{
            width: '100%',
            height: '100vh',
            backgroundColor: '#00000099',
            position: 'absolute',
          }}
          onClick={() => {
            setModal({ ...modal, show: false });
          }}
        >
          <form
            style={{
              ...positionStylingProps,
              position: 'absolute',
              backgroundColor: 'white',
              display: 'inline-block',
              padding: '20px',
              borderRadius: '8px',
              color: 'red',
              border: '1px solid black',
            }}
            onClick={(e) => e.stopPropagation()}
            onSubmit={(e) => {
              e.preventDefault();
              if (modal.onSave(modal.fields))
                setModal({ ...modal, show: false });
            }}
          >
            <h2>{modal.title}</h2>
            <br />
            {Object.entries(modal.fields).map(([key, value]) => (
              <div key={key}>
                <Input
                  label={key}
                  value={value}
                  onChange={(e) =>
                    setModal({
                      ...modal,
                      fields: { ...modal.fields, [key]: e.target.value },
                    })
                  }
                />
                <br />
              </div>
            ))}

            <center>
              <Button type='submit'>Save</Button>
            </center>
          </form>
        </div>
      )}
      {children}
    </ModalContext.Provider>
  );
}

ModalContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
