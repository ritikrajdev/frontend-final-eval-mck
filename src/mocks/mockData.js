export const mockCollections = [
  {
    id: 5,
    name: 'Laptops',
    schema: {
      brand: 'text',
    },
    userId: 1,
    formResponses: [
      {
        id: 8,
        form_id: 5,
        response: {
          brand: 'dell',
        },
      },
      {
        id: 4,
        form_id: 5,
        response: {
          brand: 'apple',
        },
      },
    ],
  },
];
