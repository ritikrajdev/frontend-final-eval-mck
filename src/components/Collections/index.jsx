import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import {
  createResponseApiEndpoint,
  editResponseApiEndpoint,
} from '../../constants/apiEndpoints';
import { ContentContext } from '../../contexts/ContentContext';
import { ErrorContext } from '../../contexts/ErrorContext';
import { ModalContext } from '../../contexts/ModalContext';
import { makeRequest } from '../../utils/makeRequest';

import './Collections.css';

export default function Collections({ collectionId }) {
  // eslint-disable-next-line no-unused-vars
  const [contentTypes, setContentTypes] = useContext(ContentContext);

  // eslint-disable-next-line no-unused-vars
  const [error, setError] = useContext(ErrorContext);

  // eslint-disable-next-line no-unused-vars
  const [modal, setModal] = useContext(ModalContext);

  const selectedCollection = contentTypes.find(
    (collection) => collection.id === collectionId
  );

  function addNewEntry() {
    const modalFields = { ...selectedCollection.schema };
    Object.keys(modalFields).forEach((key) => {
      modalFields[key] = '';
    });

    setModal({
      show: true,
      title: 'Add Response',
      position: 'right',
      fields: modalFields,
      onSave: async (data) => {
        try {
          const newResponse = await makeRequest(createResponseApiEndpoint, {
            data: {
              form_id: selectedCollection.id,
              response: data,
            },
          });

          setContentTypes(
            contentTypes.map((collection) => {
              if (collection.id === collectionId) {
                return {
                  ...collection,
                  formResponses: [...collection.formResponses, newResponse],
                };
              }
              return collection;
            })
          );

          return true;
        } catch (err) {
          setError(err.response ? err.response.data.message : err.message);
          return false;
        }
      },
    });
  }

  function editResponse(id) {
    setModal({
      show: true,
      title: 'Edit Response',
      position: 'right',
      fields: selectedCollection.formResponses.find(
        (response) => response.id === id
      ).response,
      onSave: async (data) => {
        try {
          const updatedResponse = await makeRequest(
            editResponseApiEndpoint(id),
            {
              data: {
                form_id: selectedCollection.id,
                response: data,
              },
            }
          );

          setContentTypes(
            contentTypes.map((collection) => {
              if (collection.id === collectionId) {
                return {
                  ...collection,
                  formResponses: collection.formResponses.map((response) => {
                    if (response.id === id) {
                      return updatedResponse;
                    }
                    return response;
                  }),
                };
              }
              return collection;
            })
          );

          return true;
        } catch (err) {
          setError(err.response ? err.response.data.message : err.message);
          return false;
        }
      },
    });
  }

  function deleteResponse(id) {
    // TODO
    console.log('delete response', id);
  }

  const schema = selectedCollection.schema;
  const requiredKeys = Object.keys(schema).slice(0, 3);
  requiredKeys.unshift('id');

  return (
    <div className='collections'>
      <h2>{selectedCollection.name}</h2>
      <br />
      <div
        style={{
          padding: '20px',
        }}
      >
        <div className='entries'>
          <h1>{selectedCollection.formResponses.length} Entries Found</h1>
          <button onClick={addNewEntry}>Add a new entry</button>
        </div>

        <div className='entries-table'>
          <div className='entries-headings'>
            {requiredKeys.map((key) => (
              <div key={key} className='w-150'>
                {key}
              </div>
            ))}
            <div
              style={{
                marginLeft: 'auto',
              }}
            >
              Actions
            </div>
          </div>

          {selectedCollection.formResponses.map((entry) => (
            <div key={entry.id} className='entries-row'>
              <div className='w-150'>{entry.id}</div>
              {requiredKeys.slice(1).map((key) => (
                <div key={key} className='w-150'>
                  {entry.response[key]}
                </div>
              ))}
              <div style={{ marginLeft: 'auto', display: 'flex', gap: '12px' }}>
                {/* <button>
                  <i className='fa-regular fa-clone'></i>
                </button> */}

                <button onClick={() => editResponse(entry.id)}>
                  <i className='fa-regular fa-pen-to-square'></i>
                </button>

                <button onClick={() => deleteResponse(entry.id)}>
                  <i className='fa-regular fa-trash-can'></i>
                </button>
              </div>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
}

Collections.propTypes = {
  collectionId: PropTypes.number.isRequired,
};
