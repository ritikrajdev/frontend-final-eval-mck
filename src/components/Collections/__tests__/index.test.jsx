import { render } from '@testing-library/react';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import Collections from '..';
import { ContentContext } from '../../../contexts/ContentContext';
import { ErrorContext } from '../../../contexts/ErrorContext';
import { mockCollections } from '../../../mocks/mockData';

jest.mock('react-router-dom');

describe('Collections', () => {
  it('should render correctly', async () => {
    const navigate = jest.fn();
    useNavigate.mockReturnValue(navigate);

    const error = 'some error';
    const setError = jest.fn();

    const collections = mockCollections;
    const setCollections = jest.fn();

    const { asFragment } = render(
      <ErrorContext.Provider value={[error, setError]}>
        <ContentContext.Provider value={[collections, setCollections]}>
          <Collections collectionId={mockCollections[0].id} />
        </ContentContext.Provider>
      </ErrorContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
