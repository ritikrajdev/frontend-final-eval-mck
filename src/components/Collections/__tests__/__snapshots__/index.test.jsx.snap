// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collections should render correctly 1`] = `
<DocumentFragment>
  <div
    class="collections"
  >
    <h2>
      <span>
        Laptops
      </span>
       
      <button
        class="sign-out-button"
      >
        Sign Out
      </button>
    </h2>
    <br />
    <div
      style="padding: 20px;"
    >
      <div
        class="entries"
      >
        <h1>
          2 Entries Found
        </h1>
        <button>
          Add a new entry
        </button>
      </div>
      <div
        class="entries-table"
      >
        <div
          class="entries-headings"
        >
          <div
            class="w-150"
          >
            id
          </div>
          <div
            class="w-150"
          >
            brand
          </div>
          <div
            style="margin-left: auto;"
          >
            Actions
          </div>
        </div>
        <div
          class="entries-row"
        >
          <div
            class="w-150"
          >
            8
          </div>
          <div
            class="w-150"
          >
            dell
          </div>
          <div
            style="margin-left: auto; display: flex; gap: 12px;"
          >
            <button>
              <i
                class="fa-regular fa-pen-to-square"
              />
            </button>
            <button>
              <i
                class="fa-regular fa-trash-can"
              />
            </button>
          </div>
        </div>
        <div
          class="entries-row"
        >
          <div
            class="w-150"
          >
            4
          </div>
          <div
            class="w-150"
          >
            apple
          </div>
          <div
            style="margin-left: auto; display: flex; gap: 12px;"
          >
            <button>
              <i
                class="fa-regular fa-pen-to-square"
              />
            </button>
            <button>
              <i
                class="fa-regular fa-trash-can"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Collections should render correctly after clicking on type button 1`] = `
<DocumentFragment>
  <div
    class="collections"
  >
    <h2>
      <span>
        Laptops
      </span>
       
      <button
        class="sign-out-button"
      >
        Sign Out
      </button>
    </h2>
    <br />
    <div
      style="padding: 20px;"
    >
      <div
        class="entries"
      >
        <h1>
          2 Entries Found
        </h1>
        <button>
          Add a new entry
        </button>
      </div>
      <div
        class="entries-table"
      >
        <div
          class="entries-headings"
        >
          <div
            class="w-150"
          >
            id
          </div>
          <div
            class="w-150"
          >
            brand
          </div>
          <div
            style="margin-left: auto;"
          >
            Actions
          </div>
        </div>
        <div
          class="entries-row"
        >
          <div
            class="w-150"
          >
            8
          </div>
          <div
            class="w-150"
          >
            dell
          </div>
          <div
            style="margin-left: auto; display: flex; gap: 12px;"
          >
            <button>
              <i
                class="fa-regular fa-pen-to-square"
              />
            </button>
            <button>
              <i
                class="fa-regular fa-trash-can"
              />
            </button>
          </div>
        </div>
        <div
          class="entries-row"
        >
          <div
            class="w-150"
          >
            4
          </div>
          <div
            class="w-150"
          >
            apple
          </div>
          <div
            style="margin-left: auto; display: flex; gap: 12px;"
          >
            <button>
              <i
                class="fa-regular fa-pen-to-square"
              />
            </button>
            <button>
              <i
                class="fa-regular fa-trash-can"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
